### PR TITLE
mgr/dashboard: make application field for pool creation mandatory

### PR DESCRIPTION
**Pull request by avanthakkar**
Fixes: https://tracker.ceph.com/issues/61238

Before
Screenshot from 2023-05-18 18-06-15

After:
Screenshot from 2023-05-19 15-10-21

Screen cast:
[Screencast from 2024-06-18 12-41-02.webm](https://github.com/ceph/ceph/assets/23611106/7bfcb4bf-7d59-43df-8b72-f43067f6575f)

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by pereman2 — CHANGES_REQUESTED** [review]
Didn't work. It created the pool after saying it was required.

[screen-capture (8).webm](https://github.com/ceph/ceph/assets/30913090/80deccf5-232f-4b74-9e30-1b8011b23f35)

**Review by Pegonzal — APPROVED** [review]
Just a NIT

The: `This field is required!` text only shows if all other fields have been filled

image

image

**Comment by avanthakkar** [comment]
jenkins test make check

**Review by pereman2 — APPROVED** [review]

**Review by nizamial09 — COMMENTED** [review]
Probably it should be better to add a helper here saying that a pool cannot be used without [associating with an application before](https://docs.ceph.com/en/latest/rados/operations/pools/#associate-pool-to-application)

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form-data.ts [hunk 1]
```diff
@@ -17,7 +17,7 @@ export class PoolFormData {
       available: [], // Filled during runtime
       validators: [Validators.pattern('[A-Za-z0-9_]+'), Validators.maxLength(128)],
       messages: new SelectMessages({
-        empty: $localize`No applications added`,
+        empty: '',
```
[the remark this answers is not in the document]
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form-data.ts]: Added the empty message!

**Comment by avanthakkar** [comment]
> Probably it should be better to add a helper here saying that a pool cannot be used without [associating with an application before](https://docs.ceph.com/en/latest/rados/operations/pools/#associate-pool-to-application)

And added a helper for application fields.

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
make-check failure
```
● PoolFormComponent › submit - create › erasure coded pool › minimum requirements without ECP to create ec pool

    expect(spy).toHaveBeenCalledWith(...expected)

    Expected: {"pg_autoscale_mode": "off", "pg_num": 4, "pool": "minECPool", "pool_type": "erasure"}

    Number of calls: 0

      82 |     spyOn(taskWrapper, 'wrapTaskAroundCall').and.callThrough();
      83 |     component.submit();
    > 84 |     expect(poolService[poolServiceMethod]).toHaveBeenCalledWith(pool);
         |                                            ^
      85 |     expect(taskWrapper.wrapTaskAroundCall).toHaveBeenCalledWith({
      86 |       task: {
      87 |         name: taskName,

      at expectValidSubmit (src/app/ceph/pool/pool-form/pool-form.component.spec.ts:84:44)
      at src/app/ceph/pool/pool-form/pool-form.component.spec.ts:1108:9
      at _ZoneDelegate.Object.<anonymous>._ZoneDelegate.invoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:409:30)
      at ProxyZoneSpec.Object.<anonymous>.ProxyZoneSpec.onInvoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:3830:43)
      at _ZoneDelegate.Object.<anonymous>._ZoneDelegate.invoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:408:56)
      at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:169:47)
      at Object.wrappedFunc (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:4330:34)

  ● PoolFormComponent › submit - create › erasure coded pool › creates ec pool with erasure coded profile

    expect(spy).toHaveBeenCalledWith(...expected)

    - Expected
    + Received

    @@ -1,6 +1,10 @@
      Object {
    +   "application_metadata": Array [
    +     "cephfs",
    +     "rgw",
    +   ],
        "erasure_code_profile": "ecpMinimalMock",
        "pg_autoscale_mode": "off",
        "pg_num": 4,
        "pool": "ecPool",
        "pool_type": "erasure",,

    Number of calls: 1

      82 |     spyOn(taskWrapper, 'wrapTaskAroundCall').and.callThrough();
      83 |     component.submit();
    > 84 |     expect(poolService[poolServiceMethod]).toHaveBeenCalledWith(pool);
         |                                            ^
      85 |     expect(taskWrapper.wrapTaskAroundCall).toHaveBeenCalledWith({
      86 |       task: {
      87 |         name: taskName,

      at expectValidSubmit (src/app/ceph/pool/pool-form/pool-form.component.spec.ts:84:44)
      at expectEcSubmit (src/app/ceph/pool/pool-form/pool-form.component.spec.ts:1077:9)
      at src/app/ceph/pool/pool-form/pool-form.component.spec.ts:1122:9
      at _ZoneDelegate.Object.<anonymous>._ZoneDelegate.invoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:409:30)
      at ProxyZoneSpec.Object.<anonymous>.ProxyZoneSpec.onInvoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:3830:43)
      at _ZoneDelegate.Object.<anonymous>._ZoneDelegate.invoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:408:56)
      at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:169:47)
      at Object.wrappedFunc (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:4330:34)

  ● PoolFormComponent › submit - create › erasure coded pool › creates ec pool with ec_overwrite flag

    expect(spy).toHaveBeenCalledWith(...expected)

    - Expected
    + Received

    @@ -1,6 +1,10 @@
      Object {
    +   "application_metadata": Array [
    +     "cephfs",
    +     "rgw",
    +   ],
        "erasure_code_profile": "ecp1",
        "flags": Array [
          "ec_overwrites",
        ],
        "pg_autoscale_mode": "off",,

    Number of calls: 1

      82 |     spyOn(taskWrapper, 'wrapTaskAroundCall').and.callThrough();
      83 |     component.submit();
    > 84 |     expect(poolService[poolServiceMethod]).toHaveBeenCalledWith(pool);
         |                                            ^
      85 |     expect(taskWrapper.wrapTaskAroundCall).toHaveBeenCalledWith({
      86 |       task: {
      87 |         name: taskName,

      at expectValidSubmit (src/app/ceph/pool/pool-form/pool-form.component.spec.ts:84:44)
      at expectEcSubmit (src/app/ceph/pool/pool-form/pool-form.component.spec.ts:1077:9)
      at src/app/ceph/pool/pool-form/pool-form.component.spec.ts:1132:9
      at _ZoneDelegate.Object.<anonymous>._ZoneDelegate.invoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:409:30)
      at ProxyZoneSpec.Object.<anonymous>.ProxyZoneSpec.onInvoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:3830:43)
      at _ZoneDelegate.Object.<anonymous>._ZoneDelegate.invoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:408:56)
      at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:169:47)
      at Object.wrappedFunc (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:4330:34)
```

**Comment by nizamial09** [comment]
@avanthakkar ping^

**Comment by avanthakkar** [comment]
> make-check failure
> 
> ```
> ● PoolFormComponent › submit - create › erasure coded pool › minimum requirements without ECP to create ec pool
> 
>     expect(spy).toHaveBeenCalledWith(...expected)
> 
>     Expected: {"pg_autoscale_mode": "off", "pg_num": 4, "pool": "minECPool", "pool_type": "erasure"}
> 
>     Number of calls: 0
> 
>       82 |     spyOn(taskWrapper, 'wrapTaskAroundCall').and.callThrough();
>       83 |     component.submit();
>     > 84 |     expect(poolService[poolServiceMethod]).toHaveBeenCalledWith(pool);
>          |                                            ^
>       85 |     expect(taskWrapper.wrapTaskAroundCall).toHaveBeenCalledWith({
>       86 |       task: {
>       87 |         name: taskName,
> 
>       at expectValidSubmit (src/app/ceph/pool/pool-form/pool-form.component.spec.ts:84:44)
>       at src/app/ceph/pool/pool-form/pool-form.component.spec.ts:1108:9
>       at _ZoneDelegate.Object.<anonymous>._ZoneDelegate.invoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:409:30)
>       at ProxyZoneSpec.Object.<anonymous>.ProxyZoneSpec.onInvoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:3830:43)
>       at _ZoneDelegate.Object.<anonymous>._ZoneDelegate.invoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:408:56)
>       at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:169:47)
>       at Object.wrappedFunc (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:4330:34)
> 
>   ● PoolFormComponent › submit - create › erasure coded pool › creates ec pool with erasure coded profile
> 
>     expect(spy).toHaveBeenCalledWith(...expected)
> 
>     - Expected
>     + Received
> 
>     @@ -1,6 +1,10 @@
>       Object {
>     +   "application_metadata": Array [
>     +     "cephfs",
>     +     "rgw",
>     +   ],
>         "erasure_code_profile": "ecpMinimalMock",
>         "pg_autoscale_mode": "off",
>         "pg_num": 4,
>         "pool": "ecPool",
>         "pool_type": "erasure",,
> 
>     Number of calls: 1
> 
>       82 |     spyOn(taskWrapper, 'wrapTaskAroundCall').and.callThrough();
>       83 |     component.submit();
>     > 84 |     expect(poolService[poolServiceMethod]).toHaveBeenCalledWith(pool);
>          |                                            ^
>       85 |     expect(taskWrapper.wrapTaskAroundCall).toHaveBeenCalledWith({
>       86 |       task: {
>       87 |         name: taskName,
> 
>       at expectValidSubmit (src/app/ceph/pool/pool-form/pool-form.component.spec.ts:84:44)
>       at expectEcSubmit (src/app/ceph/pool/pool-form/pool-form.component.spec.ts:1077:9)
>       at src/app/ceph/pool/pool-form/pool-form.component.spec.ts:1122:9
>       at _ZoneDelegate.Object.<anonymous>._ZoneDelegate.invoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:409:30)
>       at ProxyZoneSpec.Object.<anonymous>.ProxyZoneSpec.onInvoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:3830:43)
>       at _ZoneDelegate.Object.<anonymous>._ZoneDelegate.invoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:408:56)
>       at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:169:47)
>       at Object.wrappedFunc (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:4330:34)
> 
>   ● PoolFormComponent › submit - create › erasure coded pool › creates ec pool with ec_overwrite flag
> 
>     expect(spy).toHaveBeenCalledWith(...expected)
> 
>     - Expected
>     + Received
> 
>     @@ -1,6 +1,10 @@
>       Object {
>     +   "application_metadata": Array [
>     +     "cephfs",
>     +     "rgw",
>     +   ],
>         "erasure_code_profile": "ecp1",
>         "flags": Array [
>           "ec_overwrites",
>         ],
>         "pg_autoscale_mode": "off",,
> 
>     Number of calls: 1
> 
>       82 |     spyOn(taskWrapper, 'wrapTaskAroundCall').and.callThrough();
>       83 |     component.submit();
>     > 84 |     expect(poolService[poolServiceMethod]).toHaveBeenCalledWith(pool);
>          |                                            ^
>       85 |     expect(taskWrapper.wrapTaskAroundCall).toHaveBeenCalledWith({
>       86 |       task: {
>       87 |         name: taskName,
> 
>       at expectValidSubmit (src/app/ceph/pool/pool-form/pool-form.component.spec.ts:84:44)
>       at expectEcSubmit (src/app/ceph/pool/pool-form/pool-form.component.spec.ts:1077:9)
>       at src/app/ceph/pool/pool-form/pool-form.component.spec.ts:1132:9
>       at _ZoneDelegate.Object.<anonymous>._ZoneDelegate.invoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:409:30)
>       at ProxyZoneSpec.Object.<anonymous>.ProxyZoneSpec.onInvoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:3830:43)
>       at _ZoneDelegate.Object.<anonymous>._ZoneDelegate.invoke (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:408:56)
>       at Zone.Object.<anonymous>.Zone.run (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:169:47)
>       at Object.wrappedFunc (node_modules/zone.js/bundles/zone-testing-bundle.umd.js:4330:34)
> ```

Fixed unit tests, passing locally!

**Comment by avanthakkar** [comment]
jenkins retest this please

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html [hunk 2]
```diff
@@ -171,11 +171,16 @@
         </div>
         <!-- Applications -->
         <div class="form-group row">
-          <label i18n
-                 class="cd-col-form-label"
-                 for="applications">Applications</label>
+          <label class="cd-col-form-label required"
+                 for="applications">
+            <ng-container i18n>Applications</ng-container>
+            <cd-helper>
```
Comment (afreen23) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:177]: Let's use <cd-help-text> after the input tag here as well. 
https://github.com/ceph/ceph/blob/ef9bf7651800632d10bacef3c41f3432db769f40/src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-bucket-form/rgw-bucket-form.component.html#L252-L254 

We need to gradually move to this style, for all new things added and rest will be taken care by carbon

Thanks!

**Review by afreen23 — DISMISSED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html (hunk 2 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:177]: I've already addressed this here https://github.com/ceph/ceph/pull/51591#. So once that's merged I can simple rebase this PR. Marking this PR as DNM for now and after https://github.com/ceph/ceph/pull/51591#

**Comment by avanthakkar** [comment]
After https://github.com/ceph/ceph/pull/51591#

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test make check arm64

**Comment by avanthakkar** [comment]
jenkins test dashboard

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html (hunk 2 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:177]: @afreen23 I've merged  https://github.com/ceph/ceph/pull/51591 and rebased this PR, can you review it again?

**Review by sunilangadi2 — CHANGES_REQUESTED** [review]
@avanthakkar application is not mandatory for pool creation,
why are we making it as `This field is required!`?

**Comment by avanthakkar** [comment]
> @avanthakkar application is not mandatory for pool creation, why are we making it as `This field is required!`?
Well you have to add some application anyways after creation else it'll show health_warn

**Review by sunilangadi2 — APPROVED** [review]
yes it make sense from CLI as well
```[root@magna011 ~]# ceph health detail
HEALTH_WARN 1 pool(s) do not have an application enabled; 4 pool(s) have non-power-of-two pg_num
[WRN] POOL_APP_NOT_ENABLED: 1 pool(s) do not have an application enabled
    application not enabled on pool 'rep_pool_mGnGGtgZwA'
    use 'ceph osd pool application enable <pool-name> <app-name>', where <app-name> is 'cephfs', 'rbd', 'rgw', or freeform for custom applications.```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html [hunk 3]
```diff
@@ -189,6 +194,9 @@
                title="Pools should be associated with an application tag"
                class="{{icons.warning}} icon-warning-color">
             </i>
+            <span class="invalid-feedback"
+                  *ngIf="!isApplicationsSelected"
+                  i18n>This field is required!</span>
```
Comment (cloudbehl) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/pool/pool-form/pool-form.component.html:199]:
```suggestion
                  i18n>Application selection is required!</span>
```

**Comment by cloudbehl** [comment]
The validation sticks around even after selecting the application. Can you make sure once user selects the application this goes away.
[Screencast from 2024-05-30 11-17-14.webm](https://github.com/ceph/ceph/assets/6695156/c72c24b7-b46b-41ac-89bb-046ec58698fd)

**Comment by avanthakkar** [comment]
> The validation sticks around even after selecting the application. Can you make sure once user selects the application this goes away.
> Screencast.from.2024-05-30.11-17-14.webm

Fixed the issue & attached new screencast in PR description!

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by avanthakkar** [comment]
jenkins test dashboard

**Comment by avanthakkar** [comment]
jenkins test dashboard cephadm

**Comment by avanthakkar** [comment]
jenkins test make check arm64

**Pull request merged**